### PR TITLE
Top level values can be only accessed from the main thread.

### PR DESCRIPTION
**Pull request by olonho**

### CODE REVIEW

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt [hunk 1]
```diff
@@ -49,12 +50,27 @@ import org.jetbrains.kotlin.resolve.descriptorUtil.classId
...
+    descriptor.annotations.hasAnnotation(threadLocalAnnotationFqName)
+
 val IrField.isShared get() =
-    !descriptor.annotations.hasAnnotation(threadLocalAnnotationFqName) && !descriptor.isVar
+    descriptor.annotations.hasAnnotation(sharedAnnotationFqName)
+
+val IrField.isMainOnly get() =
+    !descriptor.annotations.hasAnnotation(threadLocalAnnotationFqName) &&
+    !descriptor.annotations.hasAnnotation(sharedAnnotationFqName) &&
+    !descriptor.isDelegated
```
Comment (olonho) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt]: Is such a check actually correct?

File: runtime/src/main/cpp/Runtime.cpp [hunk 2]
```diff
@@ -82,6 +83,7 @@ RuntimeState* initRuntime() {
   bool firstRuntime = atomicAdd(&aliveRuntimesCount, 1) == 1;
   // Keep global variables in state as well.
   if (firstRuntime) {
+    isMainThread = 1;
```
Comment (SvyatoslavScherbina) [inline runtime/src/main/cpp/Runtime.cpp]: What if Kotlin runtime gets initialized on non-main thread first?

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt [hunk 3]
```diff
@@ -49,12 +50,27 @@ import org.jetbrains.kotlin.resolve.descriptorUtil.classId
 import org.jetbrains.kotlin.resolve.descriptorUtil.fqNameUnsafe
 
 private val threadLocalAnnotationFqName = FqName("kotlin.native.ThreadLocal")
+private val sharedAnnotationFqName = FqName("kotlin.native.SharedImmutable")
 
+// TODO: maybe unannotated singleton objects shall be accessed from main thread only as well?
 val IrClass.objectIsShared get() =
     !descriptor.annotations.hasAnnotation(threadLocalAnnotationFqName)
 
+val IrField.isThreadLocal get() =
```
Comment (SvyatoslavScherbina) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt]: Are this predicates supposed to be mutually exclusive?
If so, then it makes sense to make a enum-returning function instead.

File: runtime/src/main/cpp/Runtime.cpp (hunk 2 above)
Comment (olonho) [reply runtime/src/main/cpp/Runtime.cpp]: It is impossible with this definition of the main thread - main thread *is* the first thread which initialized the runtime. It's up to platform-specific code to ensure that Kotlin/Native main thread matches to GUI thread or whatever is the the main thread in other senses.

Comment (SvyatoslavScherbina) [reply runtime/src/main/cpp/Runtime.cpp]: > It's up to platform-specific code to ensure that Kotlin/Native main thread matches to GUI thread

How do you suggest to implement this on iOS then?

Comment (olonho) [reply runtime/src/main/cpp/Runtime.cpp]: Call properly in launcher for direct interop, let's discuss how to do for reverse interop. My gut feeling is that somewhere near https://github.com/JetBrains/kotlin-native/blob/9b2fc6c6dddfa9e749eea21dce8f3fb79a34b1a4/runtime/src/main/cpp/ObjCExport.mm#L418 may help.

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt (hunk 3 above)
Comment (olonho) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt]: Hmm, not sure if they must be mutually exclusive, they are more like computations of expected policy of accessing the field.

**Pull request merged**